### PR TITLE
Internal promotion for 2021.Q4.3

### PR DESCRIPTION
**Pull request by JaxLinAMD**
internal change for xgl, pal , spvgen has been pushed to GPUOpen-Drivers. 
llpc change is created as a pull request here for google engineers to review.

### CODE REVIEW

**Comment by JaxLinAMD** [comment]
@kuhar  please take a look. other PRs may fails test because internal promotion for xgl, pal, spvgen has been pushed but internal
change for llpc is still pending here for code review.

**Comment by amdvlk-admin** [comment]
### Test summary for commit eaadab248035769134c28a471414704f7ccb97f5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 38aa59bcf134d1cb340ddd2459b34b6eda798abe</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c6950e03a4c97beadc88a8bf5bea3dc5aa2649ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225376)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32512/56344 (57.7%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23832/56344 (42.3%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32352/56344 (57.4%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23991/56344 (42.6%)</li><li>Warnings: 1/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32688/56344 (58.0%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 23656/56344 (42.0%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35669/56344 (63.3%)</li>
<li>Failed: 0/56344 (0.0%)</li>
<li>Not Supported: 20675/56344 (36.7%)</li><li>Warnings: 0/56344 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/ArithBuilder.cpp [hunk 1]
```diff
@@ -609,8 +609,7 @@ Value *ArithBuilder::CreateLog(Value *x, const Twine &instName) {
 Value *ArithBuilder::CreateSqrt(Value *x, const Twine &instName) {
   if (x->getType()->getScalarType()->isDoubleTy()) {
     // NOTE: For double type, the SQRT and RSQ instructions don't have required precision, we apply Goldschmidt's
-    // algorithm (https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Goldschmidt%E2%80%99s_algorithm) to
-    // improve the result:
+    // algorithm to improve the result:
```
Comment (kuhar) [inline lgc/builder/ArithBuilder.cpp:612]: Why was this removed?

File: lgc/include/lgc/patch/PatchLoopMetadata.h [hunk 2]
```diff
@@ -33,7 +33,6 @@
 #include "lgc/patch/Patch.h"
 #include "lgc/state/PipelineState.h"
 #include "lgc/state/TargetInfo.h"
-//#include "llvm/Analysis/LoopAnalysisManager.h"
```
Comment (kuhar) [inline lgc/include/lgc/patch/PatchLoopMetadata.h:36]: Why was this commented out but not removed?

File: lgc/include/lgc/patch/PatchWaveSizeAdjust.h [hunk 3]
```diff
@@ -45,6 +45,8 @@ class PatchWaveSizeAdjust final : public llvm::PassInfoMixin<PatchWaveSizeAdjust
   bool runImpl(llvm::Module &module, PipelineState *pipelineState);
 
   static llvm::StringRef name() { return "Patch LLVM for per-shader wave size adjustment"; }
+
+private:
```
Comment (kuhar) [inline lgc/include/lgc/patch/PatchWaveSizeAdjust.h:49]: Why do we need this?

File: llpc/lower/llpcSpirvLowerGlobal.cpp [hunk 4]
```diff
@@ -1031,7 +1031,8 @@ Value *SpirvLowerGlobal::addCallInstForInOutImport(Type *inOutTy, unsigned addrS
             assert(inOutMeta.InterpLoc == InterpLocCustom);
             elem = addCallInstForInOutImport(elemTy, addrSpace, elemMeta, nullptr, 0, nullptr, 0, inOutMeta.InterpLoc,
                                              m_builder->getInt32(idx), true, insertPos);
-          } else {
+          }
```
Comment (kuhar) [inline llpc/lower/llpcSpirvLowerGlobal.cpp:1034]: missing `// clang-format off`?

File: llpc/tool/llpcPipelineBuilder.cpp [hunk 5]
```diff
@@ -161,13 +166,17 @@ Result PipelineBuilder::buildGraphicsPipeline(BinaryData &outBinaryData) {
...
   if (result != Result::Success)
     return result;
 
+  SmallVector<BinaryData, 1> pipelines;
   outBinaryData = pipelineOut->pipelineBin;
+  pipelines.push_back(outBinaryData);
+
+  auto onExit = make_scope_exit([&] { runPostBuildActions(pipelineDumpHandle, pipelines); });
+
   return Result::Success;
```
Comment (kuhar) [inline llpc/tool/llpcPipelineBuilder.cpp:180]: This change brakes the existing behavior. The clean up code won't be run when `BuildGraphicsPipeline` fails. And if we did want to run it just before returning, `runPostBuildActions` should be called directly.

File: llpc/tool/llpcPipelineBuilder.cpp [hunk 6]
```diff
@@ -218,13 +227,17 @@ Result PipelineBuilder::buildComputePipeline(BinaryData &outBinaryData) {
...
 
   Result result = m_compiler.BuildComputePipeline(pipelineInfo, pipelineOut, pipelineDumpHandle);
   if (result != Result::Success)
     return result;
 
   outBinaryData = pipelineOut->pipelineBin;
+  SmallVector<BinaryData, 1> pipelines;
+  pipelines.push_back(outBinaryData);
+
+  auto onExit = make_scope_exit([&] { runPostBuildActions(pipelineDumpHandle, pipelines); });
```
Comment (kuhar) [inline llpc/tool/llpcPipelineBuilder.cpp:239]: Same here.

File: util/vkgcElfReader.h [hunk 7]
```diff
@@ -392,7 +392,7 @@ struct MsgPackIterator {
 };
 
 // =====================================================================================================================
-// Represents a reader for loading data from an [Executable and Linkable Format (ELF)](http://tinyurl.com/2toj8) buffer.
+// Represents a reader for loading data from an [Executable and Linkable Format (ELF)] buffer.
```
Comment (kuhar) [inline util/vkgcElfReader.h:395]: This broke the link

File: llpc/tool/llpcPipelineBuilder.cpp (hunk 5 above)
Comment (jiaolu) [reply llpc/tool/llpcPipelineBuilder.cpp:180]: i think we can just move the code like this.

  PipelineBuildInfo localPipelineInfo = {};
  localPipelineInfo.pGraphicsInfo = pipelineInfo;
  void *pipelineDumpHandle = runPreBuildActions(localPipelineInfo);
  SmallVector<BinaryData, 1> pipelines;
  auto onExit = make_scope_exit([&] { runPostBuildActions(pipelineDumpHandle, pipelines); });

  Result result = m_compiler.BuildGraphicsPipeline(pipelineInfo, pipelineOut, pipelineDumpHandle);
  if (result != Result::Success)
    return result;

  pipelines.push_back(pipelineOut->pipelineBin);

Comment (JaxLinAMD) [reply llpc/tool/llpcPipelineBuilder.cpp:180]: @kuhar  what do you think about jiaolu's suggestion?

Comment (kuhar) [reply llpc/tool/llpcPipelineBuilder.cpp:180]: @jiaolu @JaxLinAMD This would restore the original behavior, which seems right to me. However, I think we should consider a less fragile solution that would allow you to work on multi-pipeline compilation internally, without changing the existing code around something that does not exist in the public. I made a suggestion in the 'Extends to multiple BinaryDatas' thread on our mailing list. Please reply what you think there, @jiaolu.

File: util/vkgcElfReader.h (hunk 7 above)
Comment (kuhar) [reply util/vkgcElfReader.h:395]: @JaxLinAMD If you want to properly remove the markdown reference, you should also remove the `[` `]` surrounding `Executable and Linkable Format (ELF)`

File: lgc/include/lgc/patch/PatchLoopMetadata.h (hunk 2 above)
Comment (JaxLinAMD) [reply lgc/include/lgc/patch/PatchLoopMetadata.h:36]: not sure about the reason but since it is commented, it's better to remove it.

File: lgc/include/lgc/patch/PatchWaveSizeAdjust.h (hunk 3 above)
Comment (JaxLinAMD) [reply lgc/include/lgc/patch/PatchWaveSizeAdjust.h:49]: this is caused by internal sanitization. will remove it.

File: llpc/lower/llpcSpirvLowerGlobal.cpp (hunk 4 above)
Comment (JaxLinAMD) [reply llpc/lower/llpcSpirvLowerGlobal.cpp:1034]: it should have been fixed already.

File: util/vkgcElfReader.h (hunk 7 above)
Comment (JaxLinAMD) [reply util/vkgcElfReader.h:395]: will do it
